03-debugging-w-chatgpt/debugging-w-chatgpt.md

```diff
@@ -0,0 +1,491 @@
+
+If we have trouble finding resources about the error or understanding what the code is doing using our prior skills, we can begin to augment our investigation by folding in ChatGPT to see if it can explain an error and common causes, or talk through what particular lines of code are doing in reality vs expectation. 
+
+Before applying any suggestions from generative AI tools we need to break out our information validation skills to check if what they are suggesting is accurate for the language we’re working in and addresses the intended issue. And just like any situation where we’re bug fixing or refactoring, we should be running our tests frequently and expanding our test suite if necessary to ensure that we are truly addressing the issue without causing side effects.
+
+The more we work with code, the more familiar we will be with common edge cases and coming up with the edge cases unique to our project, but we can also ask ChatGPT to help us brainstorm test cases to help strengthen our test suites. 
+
+## Check for Understanding
+
+<!-- Question 1 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: ordering
+* id: f22a6178-953d-11ee-b9d1-0242ac120002
+* title: Ordering Althea's debugging steps
+##### !question
+
+Althea is debugging another project that is producing a `TypeError` and they are trying to pinpoint the line causing them trouble. 
+Order the debugging steps that they took from first to last action:
+
+##### !end-question
+##### !answer
+
+1. Do a web search for the error
+1. Go line by line and rubber duck
+1. Use a debugger to step through the code
+1. Prompt ChatGPT for assistance pinpointing the bug
+1. Step through the code to confirm if the line ChatGPT suggested is where unexpected behavior starts to appear
+
+##### !end-answer
+##### !explanation
+
+Althea wants to first do some quick searches for the error to see if there are resources on what the error means and why it might be happening. They want to understand the code better, so they do a close reading of the code and do some rubber ducking. Once they feel more clear on the overall goal of the code, they set some breakpoints and step through the code line by line to try to suss out where the strange behavior starts. Althea has an idea of what is going wrong and has narrowed down the issue to a few lines of code but they want more clarity, so they write up what they learned from their observations and prompt ChatGPT for guidance on where the issue might be happening. After carefully reading ChatGPT’s response to ensure it makes sense, they step through their code again, paying close attention to the line that ChatGPT called out as a potential issue and ensuring they see the the bug happening at that line.
```

Comment:
```suggestion
Althea wants to first do some quick searches for the error to see if there are resources on what the error means and why it might be happening. They want to understand the code better, so they do a close reading of the code and do some rubber ducking. Once they feel more clear on the overall goal of the code, they set some breakpoints and step through the code line by line to try to suss out where the strange behavior starts. Althea has an idea of what is going wrong and has narrowed down the issue to a few lines of code but they want more clarity, so they write up what they learned from their observations and prompt ChatGPT for guidance on where the issue might be happening. After carefully reading ChatGPT’s response to ensure it makes sense, they step through their code again, paying close attention to the line that ChatGPT called out as a potential issue and double check that they see the the bug happening at that line.
```